x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/ingest/HashProcessor.java

```diff
@@ -0,0 +1,227 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.ingest.ConfigurationUtils.newConfigurationException;
+
+/**
+ * A processor that hashes the contents of a field or fields using various hashing algorithms
+ */
+public final class HashProcessor extends AbstractProcessor implements Consumer<ClusterState> {
+    public static final String TYPE = "hash";
+    public static final Setting.AffixSetting<SecureString> HMAC_KEY_SETTING = SecureSetting
+        .affixKeySetting(SecurityField.setting("ingest." + TYPE) + ".", "key",
+            (key) -> SecureSetting.secureString(key, null));
+
+    private final List<String> fields;
+    private final String targetField;
+    private final Method method;
+    private final Mac mac;
+    private final byte[] salt;
+    private final boolean ignoreMissing;
+    private final AtomicBoolean consistentHashes = new AtomicBoolean(true);
+
+    HashProcessor(String tag, List<String> fields, String targetField, byte[] salt, Method method, @Nullable Mac mac,
+                  boolean ignoreMissing) {
+        super(tag);
+        this.fields = fields;
+        this.targetField = targetField;
+        this.method = method;
+        this.mac = mac;
+        this.salt = salt;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    List<String> getFields() {
+        return fields;
+    }
+
+    String getTargetField() {
+        return targetField;
+    }
+
+    byte[] getSalt() {
+        return salt;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        if (consistentHashes.get()) {
+            Map<String, String> hashedFieldValues = fields.stream().map(f -> {
+                String value = document.getFieldValue(f, String.class, ignoreMissing);
+                if (value == null && ignoreMissing) {
+                    return new Tuple<String, String>(null, null);
+                }
+                try {
+                    return new Tuple<>(f, method.hash(mac, salt, value));
+                } catch (Exception e) {
+                    throw new IllegalArgumentException("field[" + f + "] could not be hashed", e);
+                }
+            }).filter(tuple -> Objects.nonNull(tuple.v1())).collect(Collectors.toMap(Tuple::v1, Tuple::v2));
+            if (fields.size() == 1) {
+                document.setFieldValue(targetField, hashedFieldValues.values().iterator().next());
+            } else {
+                document.setFieldValue(targetField, hashedFieldValues);
+            }
+            return document;
+        } else {
+            throw new IllegalArgumentException("inconsistent hash key");
+        }
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public void accept(ClusterState clusterState) {
+        // check hash keys for consistency and unset consistentHashes flag if inconsistent
```

Comment:
`IngestService#addIngestClusterStateListener(...)` should be invoked in the factory, otherwise this method will never be invoked.

Reply:
Also in the enrich branch the enrich processor factory implements `Consumer<ClusterState>` and here the processor. This is a subtle difference, but processor instance are created when pipelines are created and discarded when a pipeline does not exist. However if in this case we discard this kind processor then we still keep a reference to this processor via the `ingestClusterStateListeners` list in `ClusterService`. In order to make this work the following changes should be made:
* A `IngestService#removeIngestClusterStateListener(...)` should be added.
* The `innerUpdatePipelines(...)` method in `IngestService` around line 569 should invoke close on processors that implement `Closeable`.
* This processor should implement `Closeable` interface and then invoke `IngestService#removeIngestClusterStateListener(this)`

Reply:
Thanks for the review comments, @martijnvg. What do you think about centralizing the consistency checking of settings in `ConsistencySettingsService`? Each hash processor would then verify that consistency of its own hash key with the service. That would eliminate the need to track the lifecycle of the processor and would also resolve the question above about failing the pipeline validation check on the elected master node. It might also have a performance benefit if multiple hash processors used the same key since the consistency check for all of them would happen only once.

Reply:
I see, that makes sense. But then `ConsistentSettingsService` should a public component that should be made accessible in `Processor.Parameters`?

Reply:
Yes, I'll make the changes so `ConsistentSettingsService` is accessible in `Processor.Parameters` and request another review pass then.